### PR TITLE
Dt3

### PR DESCRIPTION
**Pull request by sjg20**
This is version 3 of the binding initially proposed here:

https://lkml.org/lkml/2021/10/3/185

Changes are based on the feedback and what seems to be a resolution of this issue.

Also includes a few README tweaks that might be useful.

### CODE REVIEW

File: schemas/options/u-boot,config.yaml [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+required:
+  - compatible
+
+additionalProperties: false
+
+examples:
+  - |
+    options {
+      u-boot,config {
```
Comment (robherring) [inline schemas/options/u-boot,config.yaml:97]: I think this should be just 'u-boot' for the node name. You get 1 node in the hierarchy for your component.

Whatever we decide, please define as a '$nodename' property.

File: schemas/options.yaml [hunk 2]
```diff
@@ -0,0 +1,78 @@
+# SPDX-License-Identifier: GPL-2.0-only OR BSD-2-clause
+# Copyright 2021 Google LLC
+#
+# Based on the precedent set by IEEE 1275-1994 IEEE Standard for Boot
+# (Initialization Configuration) Firmware: Core Requirements and Practices
+# https://www.openfirmware.info/data/docs/of1275.pdf
+
```
Comment (robherring) [inline schemas/options.yaml:7]: This is worth being in the description, not a comment. Someday we might generate documentation from this.

File: schemas/options.yaml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  When the firmware consists of multiple parts (various U-Boot phases, TF-A,
+  OP-TEE), it is helpful that all operate the same way at runtime, regardless of
+  how they were built. This can be achieved by passing the runtime configuration
+  (e.g. 'enable UART console', 'here are your public keys') along the chain
+  through each firmware stage. It is frustrating to have to replicate a bug on
+  production firmware which does not happen on developer firmware, because they
+  are completely different builds.
+
+  The options node provides useful functionality for this. It allows the
```
Comment (robherring) [inline schemas/options.yaml:56]: quote 'options'

Comment (robherring) [reply schemas/options.yaml:56]: And probably should be '/options'

File: schemas/options.yaml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  binaries typically do not support this. The devicetree provides a more
+  structured approach in any case.
+
+properties:
+  $nodename:
+    const: options
+
+  "#address-cells": true
+  "#size-cells": true
+  ranges: true
```
Comment (robherring) [inline schemas/options.yaml:76]: How would these properties be used? Do you expect to have translatable addresses in child nodes? If so, then only 1,2 are valid.

File: schemas/options.yaml [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+additionalProperties: false
```
Comment (robherring) [inline schemas/options.yaml:78]: This won't work. From a schema perspective, nodes are also json-schema properties. You need to say only other thing allowed is child nodes:

additionalProperties:
  type: object

**Review by robherring — CHANGES_REQUESTED** [review]

File: schemas/options.yaml (hunk 2 above)
Comment (sjg20) [reply schemas/options.yaml:7]: done

File: schemas/options.yaml (hunk 3 above)
Comment (sjg20) [reply schemas/options.yaml:56]: done

File: schemas/options.yaml (hunk 4 above)
Comment (sjg20) [reply schemas/options.yaml:76]: Possibly, but I am not sure yet.

I dropped ranges, hope that is what you mean.

File: schemas/options.yaml (hunk 5 above)
Comment (sjg20) [reply schemas/options.yaml:78]: done

**Pull request merged**